dali/kernels/imgproc/convolution/laplacian_gpu.cuh

```diff
@@ -0,0 +1,318 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_LAPLACIAN_GPU_CUH_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_LAPLACIAN_GPU_CUH_
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "dali/core/dev_buffer.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/block_setup.h"
+#include "dali/kernels/common/cast.cuh"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/convolution_gpu.h"
+#include "dali/kernels/imgproc/convolution/separable_convolution_gpu.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/scratch.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+
+namespace dali {
+namespace kernels {
+
+namespace laplacian {
+
+/**
+ * @brief Computes convolution to obtain partial derivative in one of the dimensions.
+ * Convolution consits of `axes` windows, each to convolve along one dimension of the input data,
+ * where `deriv_axis`-th window is supposed to compute partial derivative along that axis,
+ * whereas the remaining windows should perform smoothing. Setup method expects
+ * a `has_smoothing` hint if all window sizes in non-derivative directions are one. If so,
+ * the smoothing convolutions can be skipped and only a single one-dimensional
+ * convolution in derivative direction is performed.
+ */
+template <typename Out, typename In, typename W, int axes, int deriv_axis, bool has_channels,
+          bool is_sequence>
+struct PartialDerivGpu {
+  using MultiDimConv = SeparableConvolutionGpu<Out, In, W, axes, has_channels, is_sequence>;
+  static constexpr int ndim = MultiDimConv::ndim;
+  static constexpr int sequence_axes = MultiDimConv::sequence_axes;
+  using SingleDimConv = ConvolutionGpu<Out, In, W, ndim, sequence_axes + deriv_axis, has_channels>;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const std::array<TensorListShape<1>, axes>& window_sizes,
+                           bool has_smoothing) {
+    has_smoothing_ = has_smoothing;
+    if (has_smoothing_) {
+      if (!multi_dim_impl_) {
+        multi_dim_impl_ = std::make_unique<MultiDimConv>();
+      }
+      return multi_dim_impl_->Setup(ctx, in_shape, window_sizes, true);
+    }
+    if (!single_dim_impl_) {
+      single_dim_impl_ = std::make_unique<SingleDimConv>();
+    }
+    return single_dim_impl_->Setup(ctx, in_shape, window_sizes[deriv_axis]);
+  }
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim> out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const std::array<TensorListView<StorageCPU, const W, 1>, axes>& windows,
```

Comment:
I find it a bit weird that it's called "PartialDeriv", but can actually calculate something else, depending on user-provided windows. I think that the utilities to fill these windows should be either a part of this class or a utility existing somewhere nearby.

Reply:
You can find the utility (LaplacianWindows) in `laplacian_params.h` that is used in CPU implementation too. The argument ``windows` is prepared by the operator and then passed to the kernel, hence the separation. As there is only limited number of possible laplacian windows, operator has a single instance of that class and computes them once, then only putting pointer in the appropriate tensorlistviews.

Reply:
I think this can be done in a follow-up, but I do think that the function that populates these windows belongs in this header or one included by it, with clear indication int the Doxygen for `Run` how to populate these windows.

Reply:
Sure, I'll move LaplacianWindows to kernels as a separate header (to be included and used by cpu and gpu operator).